hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/launcher/ContainerLaunch.java

```diff
@@ -1382,73 +1410,53 @@ public void setExitOnFailure() {
         return Collections.emptySet();
       }
       final Set<String> deps = new HashSet<>();
-      // env/whitelistedEnv dump values inside double quotes
-      boolean inDoubleQuotes = true;
-      char c;
-      int i = 0;
-      final int len = envVal.length();
-      while (i < len) {
-        c = envVal.charAt(i);
-        if (c == '"') {
-          inDoubleQuotes = !inDoubleQuotes;
-        } else if (c == '\'' && !inDoubleQuotes) {
-          i++;
-          // eat until closing simple quote
-          while (i < len) {
-            c = envVal.charAt(i);
-            if (c == '\\') {
-              i++;
-            }
-            if (c == '\'') {
-              break;
-            }
-            i++;
+      fillDeps(deps, envVal);
+      return deps;
+    }
+
+    private void fillDeps(final Set<String> deps, final String input) {
+      Matcher matcher = mainPattern.matcher(input);
+      while (matcher.find()) {
+        String variableName = matcher.group(VARIABLE_NAME);
+        if (variableName != null && !variableName.isEmpty()) {
+          deps.add(variableName);
+        }
+
+        String doubleQuoted = matcher.group(DOUBLE_QUOTED);
+        if (doubleQuoted != null) {
+          doubleQuoted =
+              singleQuotePattern.matcher(doubleQuoted).replaceAll("");
+          fillDeps(deps, doubleQuoted);
+        }
+
+        String shellExpansion = matcher.group(SHELL_EXPANSION);
+        if (shellExpansion != null && !shellExpansion.isEmpty()) {
+          if (shellExpansion.charAt(0) == '#') {
+            shellExpansion = shellExpansion.substring(1);
           }
-        } else if (c == '\\') {
-          i++;
-        } else if (c == '$') {
-          i++;
-          if (i >= len) {
-            break;
+          String[] split = shellExpansion.split("[:#%/^,@\\[\\]]", 2);
+          if (split[0].length() > 0) {
+            deps.add(split[0]);
           }
-          c = envVal.charAt(i);
-          if (c == '{') { // for ${... bash like syntax
-            i++;
-            if (i >= len) {
-              break;
-            }
-            c = envVal.charAt(i);
-            if (c == '#') { // for ${#... bash array syntax
-              i++;
-              if (i >= len) {
-                break;
+          if (split.length == 2) {
+            Matcher variableMatcher = variableNamePattern.matcher(split[1]);
+            while (variableMatcher.find()) {
+              String vName = variableMatcher.group(VARIABLE_NAME);
```

Comment:
These lines show resemblance of variable name matching. They might be brought together.

Reply:
Extracted to a function.